main/lsp/LSPPreprocessor.cc

```diff
@@ -119,121 +98,179 @@ bool LSPPreprocessor::ensureInitialized(LSPMethod method, const LSPMessage &msg)
+                return make_unique<SignatureHelpTask>(*config, id,
+                                                      move(get<unique_ptr<TextDocumentPositionParams>>(rawParams)));
+            case LSPMethod::TextDocumentReferences:
+                return make_unique<ReferencesTask>(*config, id, move(get<unique_ptr<ReferenceParams>>(rawParams)));
+            case LSPMethod::SorbetReadFile:
+                return make_unique<SorbetReadFileTask>(*config, id,
+                                                       move(get<unique_ptr<TextDocumentIdentifier>>(rawParams)));
+            case LSPMethod::Shutdown:
+                return make_unique<ShutdownTask>(*config, id);
+            case LSPMethod::SorbetError:
+                return make_unique<SorbetErrorTask>(*config, move(get<unique_ptr<SorbetErrorParams>>(rawParams)), id);
+            default:
+                return make_unique<SorbetErrorTask>(
+                    *config,
+                    make_unique<SorbetErrorParams>(
+                        (int)LSPErrorCodes::MethodNotFound,
+                        fmt::format("Unknown request method: {}", convertLSPMethodToString(method))),
+                    id);
         }
+    } else {
+        // This should be impossible.
+        Exception::raise("Message isn't a request, notification, or response.");
     }
+}
 
-    if (shouldEnqueue || shouldMerge) {
-        absl::MutexLock lock(&stateMtx);
-        if (shouldEnqueue) {
-            state.pendingRequests.push_back(move(msg));
+bool LSPPreprocessor::cancelRequest(const CancelParams &params) {
+    absl::MutexLock lock(taskQueueMutex.get());
+    auto &pendingTasks = taskQueue->pendingTasks;
+    for (auto it = pendingTasks.begin(); it != pendingTasks.end(); ++it) {
+        auto &current = **it;
+        if (current.cancel(params.id)) {
+            pendingTasks.erase(it);
+            // Now that we've removed this request, we may be able to merge more edits together.
+            mergeFileChanges();
+            return true;
         }
-        if (shouldMerge) {
-            mergeFileChanges(stateMtx, state);
+    }
+    // It's too late; we have either already processed the request or are currently processing it. Swallow cancellation
+    // and ignore.
+    return false;
+}
+
+void LSPPreprocessor::pause() {
+    absl::MutexLock lock(taskQueueMutex.get());
+    ENFORCE(!taskQueue->paused);
+    config->logger->error("Pausing");
+    taskQueue->paused = true;
+}
+
+void LSPPreprocessor::resume() {
+    absl::MutexLock lock(taskQueueMutex.get());
+    ENFORCE(taskQueue->paused);
+    config->logger->error("Resuming");
+    taskQueue->paused = false;
+}
+
+void LSPPreprocessor::exit(int exitCode) {
+    absl::MutexLock lock(taskQueueMutex.get());
+    if (!taskQueue->terminate) {
+        taskQueue->terminate = true;
+        taskQueue->errorCode = exitCode;
+    }
+}
+
+void LSPPreprocessor::preprocessAndEnqueue(unique_ptr<LSPMessage> msg) {
```

Comment:
The new `preprocessAndEnqueue` is straightforward and easy-to-grok. No method-specific logic in here at all.